homeassistant/components/yalexs_ble/config_flow.py

```diff
@@ -0,0 +1,245 @@
+from typing import Any
+
+from bleak_retry_connector import BleakError, BLEDevice
+import voluptuous as vol
+from yalexs_ble import (
+    AuthError,
+    DisconnectedError,
+    PushLock,
+    ValidatedLockConfig,
+    local_name_is_unique,
+)
+from yalexs_ble.const import YALE_MFR_ID
+
+from homeassistant import config_entries
+from homeassistant.components.bluetooth import (
+    BluetoothServiceInfoBleak,
+    async_discovered_service_info,
+)
+from homeassistant.const import CONF_ADDRESS
+from homeassistant.data_entry_flow import AbortFlow, FlowResult
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.typing import DiscoveryInfoType
+from homeassistant.loader import async_get_integration
+
+from .const import CONF_KEY, CONF_LOCAL_NAME, CONF_SLOT, DOMAIN
+from .util import async_get_service_info, human_readable_name
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def validate_lock(
+    local_name: str, device: BLEDevice, key: str, slot: int
+) -> None:
+    """Validate a lock."""
+    if len(key) != 32:
+        raise InvalidKeyFormat
+    try:
+        bytes.fromhex(key)
+    except ValueError as ex:
+        raise InvalidKeyFormat from ex
+    if not isinstance(slot, int) or slot < 0 or slot > 255:
+        raise InvalidKeyIndex
+    await PushLock(local_name, device.address, device, key, slot).validate()
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Yale Access Bluetooth."""
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize the config flow."""
+        self._discovery_info: BluetoothServiceInfoBleak | None = None
+        self._discovered_devices: dict[str, BluetoothServiceInfoBleak] = {}
+        self._lock_cfg: ValidatedLockConfig | None = None
+
+    async def async_step_bluetooth(
+        self, discovery_info: BluetoothServiceInfoBleak
+    ) -> FlowResult:
+        """Handle the bluetooth discovery step."""
+        await self.async_set_unique_id(discovery_info.address)
+        self._abort_if_unique_id_configured()
+        self.context["local_name"] = discovery_info.name
+        self._discovery_info = discovery_info
+        self.context["title_placeholders"] = {
+            "name": human_readable_name(
+                None, discovery_info.name, discovery_info.address
+            ),
+        }
+        return await self.async_step_user()
+
+    async def async_step_integration_discovery(
+        self, discovery_info: DiscoveryInfoType
+    ) -> FlowResult:
+        """Handle a discovered integration."""
+        lock_cfg = ValidatedLockConfig(
+            discovery_info["name"],
+            discovery_info["address"],
+            discovery_info["serial"],
+            discovery_info["key"],
+            discovery_info["slot"],
+        )
+        # We do not want to raise on progress as integration_discovery takes
+        # precedence over other discovery flows since we already have the keys.
```

Comment:
There is a race here.

discovery flow: Lock is discovered and we start waiting for service info
user flow: User starts config flow and enters key and slot, validation starts
discovery flow: finishes waiting for service info, aborts flow2

So once they enter the key in the user flow, the discovery flow needs to abort instead of the user flow

Reply:
https://github.com/home-assistant/core/pull/76691